sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -0,0 +1,1248 @@
+                             self.beta_1)
+        if self.beta_2 < 0 or self.beta_2 >= 1:
+            raise ValueError("beta_2 must be >= 0 and < 1, got %s" %
+                             self.beta_2)
+        if self.epsilon <= 0.0:
+            raise ValueError("epsilon must be > 0, got %s." % self.max_iter)
+
+        # raise ValueError if not registered
+        supported_activations = ['logistic', 'tanh', 'relu']
+        if self.activation not in supported_activations:
+            raise ValueError("The activation '%s' is not supported. Supported "
+                             "activations are %s." % (self.activation,
+                                                      supported_activations))
+        if self.learning_rate not in ["constant", "invscaling", "adaptive"]:
+            raise ValueError("learning rate %s is not supported. " %
+                             self.learning_rate)
+        if self.algorithm not in _STOCHASTIC_ALGOS + ["l-bfgs"]:
+            raise ValueError("The algorithm %s is not supported. " %
+                             self.algorithm)
+
+    def _fit_lbfgs(self, X, y, activations, deltas, coef_grads,
+                   intercept_grads, layer_units):
+        # Store meta information for the parameters
+        self._coef_indptr = []
+        self._intercept_indptr = []
+        start = 0
+
+        # Save sizes and indices of coefficients for faster unpacking
+        for i in range(self.n_layers_ - 1):
+            n_fan_in, n_fan_out = layer_units[i], layer_units[i + 1]
+
+            end = start + (n_fan_in * n_fan_out)
+            self._coef_indptr.append((start, end, (n_fan_in, n_fan_out)))
+            start = end
+
+        # Save sizes and indices of intercepts for faster unpacking
+        for i in range(self.n_layers_ - 1):
+            end = start + layer_units[i + 1]
+            self._intercept_indptr.append((start, end))
+            start = end
+
+        # Run LBFGS
+        packed_coef_inter = _pack(self.coefs_,
+                                  self.intercepts_)
+
+        if self.verbose is True or self.verbose >= 1:
+            iprint = 1
+        else:
+            iprint = -1
+
+        optimal_parameters, self.loss_, d = fmin_l_bfgs_b(
+            x0=packed_coef_inter,
+            func=self._loss_grad_lbfgs,
+            maxfun=self.max_iter,
+            iprint=iprint,
+            pgtol=self.tol,
+            args=(X, y, activations, deltas, coef_grads, intercept_grads))
+
+        self._unpack(optimal_parameters)
+
+    def _fit_stochastic(self, X, y, activations, deltas, coef_grads,
+                        intercept_grads, layer_units, incremental):
+        rng = check_random_state(self.random_state)
+
+        if not incremental or not hasattr(self, '_optimizer'):
+            params = self.coefs_ + self.intercepts_
+            optimizer = stochastic_optimizers[self.algorithm]
+            self._optimizer = optimizer(params, self.learning_rate_init,
```

Comment:
I would prefer an if: else: here that delegates to each optimizer, versus passing all arguments to each optimizer